docs/notebooks/openai_gym_lunar_lander.pct.py

```diff
@@ -0,0 +1,244 @@
+    video = io.open('./lunar_lander_videos/' + filename, 'r+b').read()
+    encoded = base64.b64encode(video)
+    return HTML(data='''
+            <video width="360" height="auto" alt="test" controls>
+                <source src="data:video/mp4;base64,{0}" type="video/mp4" />
+            </video>'''.format(encoded.decode('ascii'))
+    )
+
+
+# %% [markdown]
+# #### Crash
+
+# %%
+load_video("crash.mp4")
+
+# %% [markdown]
+# #### Timeout
+
+# %%
+load_video("timeout.mp4")
+
+# %% [markdown]
+# #### Flying out of bounds
+
+# %%
+load_video("out_of_bounds.mp4")
+
+# %% [markdown]
+# #### Slamming on the surface
+# This is a very common failure mode in this environment - going too fast and slamming on the surface.
+
+# %%
+load_video("slam.mp4")
+
+# %% [markdown]
+# Finally, we need a way to deal with stochasticity of the simulation. To keep our modelling simple, let's aim at finding a controller that achieves highest average reward over a fixed number of runs. This gives us a simple way of implementing the Trieste observer that isn't too sensitive to the randomness of a single simulation. Also note that we reverse the reward sign: this is because we seek the maximum reward, and Trieste approaches optimization as minimization problem.
+
+# %%
+N_RUNS = 10
+def lander_objective(x):
+    # for each point compute average reward over n_runs runs
+    all_rewards = []
+    for w in x.numpy():
+        rewards = [
+            demo_heuristic_lander(env, w)
+            for _ in range(N_RUNS)
+        ]
+        all_rewards.append(rewards)
+
+    rewards_tensor = tf.convert_to_tensor(all_rewards, dtype=tf.float64)
+    
+    # triste minimizes, and we want to maximize
+    return -1 * tf.reshape(tf.math.reduce_mean(rewards_tensor, axis=1), (-1, 1))
+
+observer = trieste.objectives.utils.mk_observer(lander_objective)
+
+# %% [markdown]
+# ### Solving the optimization problem with Trieste
+#
+# Here we do normal steps required to solve an optimization problem with Trieste: generate some initial data, create a surrogate model, define an acquisition funciton and rule, and run the optimization. Optimization step may take a few minutes to complete.
+
+# %%
+num_initial_points = 2 * search_space.dimension
+initial_query_points = search_space.sample(num_initial_points)
+initial_data = observer(initial_query_points)
+
+
+# %%
+def build_model(data):
+    variance = tf.math.reduce_variance(data.observations)
+    kernel = gpflow.kernels.RBF(variance=variance)
+    gpr = gpflow.models.GPR(data.astuple(), kernel)
+    gpflow.set_trainable(gpr.likelihood, False)
+    # Since we are running multiple simulations per observation,
+    # it is possible to estimate likelihood variance from this data for each observation.
+    # This is possible to model with VGP, as described here:
+    # https://gpflow.readthedocs.io/en/master/notebooks/advanced/varying_noise.html\
+    # In the interest of brevity we have chosen not to do it in this notebook.
```

Comment:
This is a cool demo! I hope you don't mind me jumping in with a comment: I find the description above somewhat confusing, as it seems to throw together, on the one hand, whether your likelihood models homoskedastic noise (same variance everywhere, default in GPR) vs heteroskedastic noise (what you're alluding to with the VGP model), and on the other hand, whether you have the noise variance be trainable (default) or not (you explicitly set it to non-trainable in line 192).

Reply:
agreed, how about:
"it is possible to account for variations in observation noise by using a different likelihood variance for each observation"

Reply:
Not at all, thanks @st-- and @vpicheny for clarification